### ISSUE TITLE
What should we do with `regex.sub` and `regex.subn` methods?

### ISSUE THREAD

**Issue by e-kayrakli** [chapel-lang/chapel#19079]
These methods are documented [here](https://chapel-lang.org/docs/master/modules/standard/Regex.html#Regex.regex.subn).

#### Problems with `sub`/`subn`

- Names are not self-documenting enough. ("subn" returns a tuple that includes the new string and number of substitutions)
- The order of arguments feels very wrong to me: `myRegex.sub(replacement, string)` is quite backwards. This order is inherited from Python.

#### High-level Proposal

- We add 
  - `string.replace(pattern: regex(string), replacement:string, global=true): string` and
    - `string.replaceCounting(pattern: regex(string), replacement:string, global=true): (string, int)` or
    - `string.replace(pattern: regex(string), replacement:string, out count: int, global=true): string`
- Deprecate `regex.sub` and `regex.subn`

##### Some discussion
- A method that substitutes/replaces parts of a string based on a regular expression should have the string as its receiver and not the regex. So, I think we shouldn't have this functionality on the regex object at all.
- I don't like the `replaceCounting` name as it is too wordy. Interface-wise the version with the `out` formal feels cleaner. But the callsite might look a bit clunkier. But we have `out` formals, and it's their nature.
- The argument names `pattern` and `replacement` are subject to change. They somewhat hinge on https://github.com/chapel-lang/chapel/issues/18264

**Comment by mppf** [comment]
I like the proposal and in particular appreciate that it mirrors the existing https://chapel-lang.org/docs/builtins/String.html#String.string.replace -- however I would like it to match more exactly (with `count` instead of `global`).

**Comment by bradcray** [comment]
Just to make sure I'm understanding the proposal, we'd add these as tertiary methods in the Regex module, is that right?  (I have no real concerns about using tertiary methods in this way, just wanting to make sure I understand, and that specifically, we're not making all programs include `Regex` by default.

Generally, I like using method names/signatures that are similar across different use cases (e.g., lots of types supporting `.contains()` rather than each defining their own method name for that operation), which I think is the theme behind this proposal.

**Comment by e-kayrakli** [comment]
@mppf -- good point about `count`

@bradcray -- you're right on both.
From a user's standpoint, `use Regex` will not really bring a brand new method on `string`. It will bring an overload of `string.replace` that takes a Regex argument. FWIW, I have always been skeptical about tertiary methods. However, tertiary methods that are overloads of other primary/secondary methods feel really cool.

**Comment by bradcray** [comment]
With the caveat that I'm not particularly familiar with `subn`:

* of the two replacement names proposed, I like `replaceAndCount()` better than `replaceCounting()`
* I'm missing how the count would be accepted and returned in the proposal that you're drawn to.  Maybe you're just ignoring that aspect and focusing on the names (and other arguments)?
* I think you're right that using `inout` would prevent passing in params/consts—but that also seems appropriate if the point of using this function is to get a count out
* Is the main motivation for getting away from the tuple essentially "We're passing in a count, so that seems like a convenient place to pass it back as well"?  Python has ref arguments too, though, doesn't it?  So is it telling that they didn't do this and that we should follow their lead for familiarity?

**Comment by e-kayrakli** [comment]
> I'm missing how the count would be accepted and returned in the proposal that you're drawn to. Maybe you're just ignoring that aspect and focusing on the names (and other arguments)?

Sorry about the confusion here. I intended the bottom two to return a tuple. I updated the signatures above.

> point of using this function is to get a count out

The first version [in my comment just above](https://github.com/chapel-lang/chapel/issues/19079#issuecomment-1129082642) already has a `count` argument to control maximum number of replacements to be made. I was imagining we can get clever and have only a single function with an `inout count=-1` argument. But I thought (and still do) think it would be too confusing.

> Is the main motivation for getting away from the tuple essentially "We're passing in a count, so that seems like a convenient place to pass it back as well"? Python has ref arguments too, though, doesn't it? So is it telling that they didn't do this and that we should follow their lead for familiarity?

Does python have by-ref arg passing? I don't think so. Objects are effectively passed by reference, but primitive types are always passed by value. (They call it pass by assignment, I think. IOW, arg passing has the same semantics as assignment)

**Comment by bradcray** [comment]
> Does python have by-ref arg passing?

My mistake, I think I must've been getting confused either with... Fortran?  Or possibly the referential nature of objects.

**Comment by e-kayrakli** [comment]
On a meeting we discussed the possibility of removing the functionality that subn provides altogether. Arguments for removal:

- We are going to break away from Python's precedence, anyway
- There wasn't an obvious use case for anyone
- No similar capability in C++, Swift, Rust or Julia

Argument for keeping the functionality:

- Python has it

Some other tidbits:

- I can't find a good motivation for having subn in python.
- Arkouda has `subn`, it looks like.
- `qio_regex_replace` is the runtime function we use for both `sub` and `subn`, and it returns the number of replacements. In fact, `regex.sub` calls `regex.subn` and drops the number of replacements on the floor. So the `subn` functionality is something we get for free in terms of implementation.

---

If we keep the functionality, `replaceCounting` received no love at all. We'll go with `replaceAndCount`.

---

Another alternative is to have an overload of `string.replace` with an additional `out numReplacements` argument similar to what I had in the OP.

**Comment by e-kayrakli** [comment]
I think we are almost ready to finalize this. As a reminder, the current direction is to replace `sub` and `subn` with:

```chapel
// the optional `count` is the maximum replacement limit.
// signature is the same as base `string.replace` except for the first formal
string.replace(pattern: regex(string), replacement:string, count=-1): string

// return value has the resulting string and number of replacements
string.replaceAndCount(pattern: regex(string), replacement:string): (string, int)
```

**Comment by e-kayrakli** [comment]
Alright, things aren't as easy as changing the interface here.

RE2 doesn't expose an API where you can specify number of maximum replacement you want. You either want one replacement, or as much as there are non-overlapping matches. This is currently controlled by `global=true` flag in `sub` and `subn`.

Our `string.replace` is a homegrown implementation where we can easily control the number of replacements. With the lack of this control for RE2, we can:

- change `count=-1` to `global=true` in the above design (and add it to `replaceAndCount`)
  - This'll be slightly different than `string.replace`, but if you don't pass any actual to defaulted arguments, it should be easily swappable.
  - This'll make the change "less breaking", but different than our consensus
- keep `count=-1` (and add it to `replaceAndCount`). when there's a max count given by the user, call `RE2::Replace`'s wrapper repeatedly until you reach the max count or no more matches.
  - Similar to `string.replace`
  - More functionality as compared to the current `sub`/`subn`
  - No clear way of implementing `count!=-1` in a performant way
    - RE2 allows you to start replacement at a given index but it doesn't expose the index at which the replacement has been made. So, the implementation would arguably need to call `RE2::Replace` repeatedly *from the beginning of the string*, or we can spend more time in implementing a combo of finding the regex, doing the replacement at the found index, finding the regex again but from on offset from that previous index and so on. The latter is obviously better, but probably still slower than a single RE2 call.

Any thoughts?

**Comment by mppf** [comment]
Another option (that maybe nobody will like, but, I'm putting it out there) is for the regex versions of these methods to have the same interface as the string ones, but they would throw an error saying that the regular expression versions currently only work with `count==1` or `count==-1`.

**Comment by lydia-duncan** [comment]
I think I would say `all` instead of `global` as the name for the flag, but am otherwise okay with option 1

**Comment by lydia-duncan** [comment]
That's good context, and I suspect you are right that is the reason we use the name today.  My main objection to global is that it implies to me that it will change things beyond the string you are working on, maybe to the extent of everything involved in the program.  For some reason `all` doesn't imply the broader scope.

I don't particularly find regular expression decisions tuned for understandability, but I recognize there is value in consistency here, so I wouldn't block choosing that name for it.

**Comment by bradcray** [comment]
FWIW, before reading Lydia's comment, I had the _identical_ "Should `global` be `all`?" reaction.  The precedent is high, and I wouldn't object to `global` on those grounds.  But I don't find it very intuitive, and don't think of Perl, vim, and sed as being the gold standard we should strive for.  So that makes me wonder more about what Python does, as well as Swift and Julia (as examples of languages that seem to feel free to rename things that they think are non-intuitive).

**Comment by e-kayrakli** [comment]
> and don't think of Perl, vim, and sed as being the gold standard we should strive for.

I agree.

---

- Python's [`sub`/`subn`](https://docs.python.org/3/library/re.html#re.sub) has `count` argument. Which is option 2 above.
- Julia's [`Base.replace`](https://docs.julialang.org/en/v1/base/collections/#Base.replace-Tuple{Any,%20Vararg{Pair}}) has a `count` argument and not a `bool` flag. But this looks like a very general function and I am not sure if you can replace a string based on a regex using that. In any case, if anything, this is a datapoint pushing us towards 2 above.
- Swift doesn't have a `replace` similar to ours. It has `replaceSubrange`, which is irrelevant here. They do have `remove` and [`removeAll`](https://developer.apple.com/documentation/swift/string/removeall(where:)) to remove characters from a string matching a criterion. Which is I guess a vote towards `all`.

**Comment by bradcray** [comment]
Thanks for looking into those alternatives approaches, Engin!

I'm not enough of a regex user to have a very informed opinion between your options.  Despite that, I'm going to capture a lot of thoughts here, none of which I'm particularly firm on (which is to say, there are directions I like, but I'm not the target audience, so ...)

I feel hard-pressed to come up with cases where I wouldn't want to replace all or one instance of a substring, but since other libraries have, I also don't want to preclude it if we can help it.

I often wonder whether we'll want to support a different regex implementation someday, or the option to switch between multiple underlying back-end regex frameworks—particularly when I look at our performance on the [CLBG regex-redux](https://benchmarksgame-team.pages.debian.net/benchmarksgame/performance/regexredux.html) where we're currently 4.6x off the best entries (some significant part of which we consider to be due to RE2, I believe?).  I know when we've talked about supporting such flexibility in the past, and in those conversations it's sounded like it may be somewhere between very painful and intractable; but if someone came to us loving Chapel, yet needing us to have a regex solution that was closer in performance to C/C++/Regex, that'd be pretty motivating to do something beyond RE2 (maybe just another user-level library rather than something as baked in as RE2 is?).

That might suggest taking the @mppf approach of supporting a general count but just complaining if it wasn't one of the values we supported to support alternate implementations with the same interface in the future.  Maybe we could even support param overloads for `1` (and `-1`?) and make it a compiler error when other values are passed in (where I'm positing nobody will ever pass in a non-param count, particularly if none of them work...).

But, what I think I like better is moving away from the `-1` sentinel and towards a more Swift-like interface by having:

```chapel
string.replaceAll(pattern: regex(string), replacement: string): string  // avoids having to pass in -1 or rely on remembering what the default for `replace()` is
string.replace(pattern: regex(string), replacement:string): string      // supports adding a count argument anytime later
```

(quick rationale: It feels a little like the options available in most applications that support find...replace?)

However, given your default above, I suspect it would be unusual relative to other languages to have `replace()` just replace one match rather than all?  If so, an alternative would be to support `replaceAll()` and `replaceOne()`, leaving ourselves the opportunity to add a `replace()` that takes a count later without stepping on either of those signatures.

One other intriguing thing about using two routines here is that `replaceAll()` could take an optional `out` count argument (and `replaceOne()` wouldn't need to... unless we wanted to return 0/false or 1/true to say whether a match was found or not?).  So we wouldn't need your `replaceAndCount()`, at least until some future when `replace()` takes a `count` argument.

**Comment by e-kayrakli** [comment]
I wrote the response below the line but taking a step back, did we steer away from [Option 2 some comments above](https://github.com/chapel-lang/chapel/issues/19079#issuecomment-1419686280) because of my reluctance toward it? I do not want to over-state my negativity there, and it'd be unfortunate if that causes to over-engineer a design here. In terms of interface, that's still my favorite. `count=-1` and `count=1` cases are most likely going to be as fast as what we do today, and I just roll up my sleeves to support non-ideal `count` values. I don't think performance will be off-the-cliff if I implement the O(n) algorithm, which is not more complicated than some of the other string support we have. 

---


This all makes sense from Regex perspective. But towards the last part of your suggestion we lose symmetry with `string.replace(pattern: string, replacement: string, count=-1)`. Do you think that's OK?

I have some reservations about it. We are making a method that's pretty close to `string.replace`, but not really the same. So, I would like to consider a middle ground where we are consistent. Options I can think of:

1. Do what @mppf suggested: add `count` argument, but error if different than -1 or 1.
2. Add `string.replace(regex, string)` and `string.replaceOne(regex, string)` where the former replaces all matches and is consistent with current `string.replace` with default `count`. But then, do we need `replaceAndCount` and `replaceAndCountOne`?
3. Do what you suggested in the last paragraph above, but also add `string.replaceOne` and `string.replaceAll` on top of `string.replace`.

I am drawn more and more towards @mppf's suggestion, i.e., Option 1 above. It is pretty consistent with string interface. The cases where we'd error with illegal `count` are already not supported by the regex interface today, so no functionality is lost there. And when someone asks us to support, say, `count=3` we implement the support, even if it is not performing as fast as full replacement. (I think we should definitely avoid O(n**2) algorithm there, and O(n) isn't nightmarish to implement). The only question in my mind is how to error in the non-param case: should it throw, changing the function interface? should it halt, going against the "standard library shouldn't halt" principle? But maybe what you're suggesting is we *only* have param version until we support arbitrary counts. Which I can get behind.

**Comment by e-kayrakli** [comment]
So, based on the silence and my wish to make progress, I bit the bullet and implemented a slow(-er) version that can cover all the cases as a fall back. This is a very rudimentary test, but replacing 40 3-byte sequences with a 1-byte sequence in about 40MB of a string is only 10% slower with this new slower implementation. Note that we still use the fast version if `count==-1` (the default) or `count=1`. IOW, for the existing interface there is no performance hit. This 10% slower performance is only for new functionality. This doesn't concern me. I was expecting a bigger hit.

I pushed my current branch and opened a draft PR: https://github.com/chapel-lang/chapel/pull/21600.

@bradcray -- I know you mused about a different interface, but I had some reservations as I mentioned above. Based on your disclaimer that you're not the target audience for this, I was assuming that you're still amenable to the [original proposal here](https://github.com/chapel-lang/chapel/issues/19079#issuecomment-1374253138). Let me know if I mis-internalized your comments, and would still like to investigate a different interface.

**Comment by bradcray** [comment]
I'm definitely happy to defer to those who actually use or are familiar with other regex interfaces on this.

**Comment by e-kayrakli** [comment]
Based on already having reached consensus on the design, I went ahead and tested #21600, it all looks good. I am proceeding with the implementation in that PR.
